### PR TITLE
Support the latest pub_semver

### PR DESCRIPTION
**Pull request by nex3**
This rewrites failures' incompatibility graphs so that they refer to
real version numbers rather than the pessimistic ranges that
pub_semver can now generate.

Note that this is expected to have one failing test until
dart-lang/pub_semver#30 lands and pub_semver 1.4.1 is released.

### CODE REVIEW

File: lib/src/solver/reformat_ranges.dart [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  var terms = incompatibility.terms.map((term) {
+    var versions = packageListers[term.package.toRef()]?.cachedVersions ?? [];
+
+    if (term.package.constraint is! VersionRange) return term;
+    var range = term.package.constraint as VersionRange;
+
+    var min = range.min;
+    if (min != null && range.includeMin && min.isFirstPreRelease) {
```
Comment (natebosch) [inline lib/src/solver/reformat_ranges.dart]: We could avoid some of the reassignments if we pull out some methods.

`var min = _normalizedRangeMin(range, versions);`

File: lib/src/solver/reformat_ranges.dart [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    var minIsPreReleaseOfMax =
+        min != null && min.isPreRelease && equalsIgnoringPreRelease(min, max);
+    if (max != null &&
+        !includeMax &&
+        !max.isPreRelease &&
+        !minIsPreReleaseOfMax) {
+      var index = _lowerBound(versions, max);
+      var previous = index == 0 ? null : versions[index - 1].version;
+
+      print("original range: $range, previous version: $previous");
```
Comment (natebosch) [inline lib/src/solver/reformat_ranges.dart]: is this run inside some zone that will capture print calls?

File: lib/src/solver/reformat_ranges.dart [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            min: min,
+            max: max,
+            includeMin: range.includeMin,
+            includeMax: includeMax,
+            alwaysIncludeMaxPreRelease: true)),
+        term.isPositive);
+  });
+
+  return new Incompatibility(terms.toList(), cause);
+}
```
[the remark this answers is not in the document]
Comment (munificent) [reply lib/src/solver/reformat_ranges.dart]: +1

File: lib/src/solver/reformat_ranges.dart [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+/// Returns the first index in [ids] (which is sorted by version) whose version
+/// is greater than or equal to [version].
+///
+/// Returns `ids.length` if all the versions in `ids` are less than [version].
+///
+/// We can't use the `collection` package's `lowerBound()` function here because
+/// [version] isn't the same as [ids]' element type.
+PackageVersion _lowerBound(List<PackageId> ids, Version version) {
```
Comment (munificent) [inline lib/src/solver/reformat_ranges.dart]: The return type should be `int`, shouldn't it?

**Pull request merged**